rfc/egress-billing.md

```diff
@@ -0,0 +1,309 @@
+
+```mermaid
+sequenceDiagram
+  Alice->>Blob Service: Please store bafk...7fi
+  Blob Service-->>Alice: Stored - you can present bafy...site to fetch bafk...7fi
+  Alice-)Gateway: You can present bafy...auth to fetch bafk...7fi
+```
+
+#### Rate Limit Accounting
+
+When the gateway receives a request, it will query the [content claims service] for claims about the requested
+CID [TODO: and also other information like origin and token?]. If multiple claims establishing different rate limits
+are found, one is chosen randomly and its request count is incremented, resulting in billing to the 
+party that authorized the rate limit. If no claims are found, the public rate limit request count is incremented.
+
+These queries and the updates to rate limits they are used to create should not be done in the hot-path of reads - instead
+this should be done in background jobs and the read logic should read from whatever data store these tasks are using to manage
+rate limits.
+
+```mermaid
+sequenceDiagram
+  Gateway-)Accounting Service: I received a request for bafk...7fi
+  Accounting Service->>Content Claims Service: Give me claims relevant to this request
+  Content Claims Service-->>Accounting Service: Here are the claims I can find
+  Accounting Service->>Rate Limiting Service: Increment rate limit usage for this claim
+  Rate Limiting Service-->>Accounting Service: OK, done
+```
+
+#### Bulk Configuration
+
+NOTE: The following requires support for "delegations without inline proofs" as specified in UCAN 1.0
+
+If users would like to set rate limits on a set of CIDs in a single operation, they can create an intermediary "principle"
+to which they can address delegations - in this example we will call this principle `did:key:zGroup`. 
+
+We first create a content commitment addressed to the group:
+
+```js
+{ // bafy..group
+  "iss": "did:key:zAlice",
+  "aud": "did:key:zGroup",
+  "exp": 1716235987 // restrict to a month
+  "cmd": "/assert/location",
+  // Subject is storage node because they own that resource
+  "sub": "did:web:asia.web3.storage",
+  "pol": [
+    // Request origin header must be example.com
+    ["==", ".headers['origin']", "example.com"],
+    ["==", ".query.token", "zrptvx"]
+  ],  
+  "meta": { "proof": { "/": "bafy...site" } }
+}
+```
+
+and then create an open-ended delegation from the group to the gateway, with the same `pol` (TODO: or more restrictive?) fields:
+
+```js
+{ // bafy..access
+  "iss": "did:key:zGroup",
+  "aud": "did:web:w3s.link",
+  "exp": 1716235987 // restrict to a month
+  "cmd": "/assert/location",
+  // Subject is storage node because they own that resource
+  "sub": "did:web:asia.web3.storage",
+  "pol": [
+    // Request origin header must be example.com
+    ["==", ".headers['origin']", "example.com"],
+    ["==", ".query.token", "zrptvx"]
+  ]
+}
+```
+
+For any given CID, the gateway will be able to look up the CID-specific content commitment as well as the
```

Comment:
oh that's a good point, I actually don't know how this will handle CIDs that aren't DAG roots - do those even get content commitments back from the blob upload service? @Gozala any thoughts on this? I'm not sure we accounted for these CIDs in our chats...

Reply:
^^^ important point. I think there's an intersection where we need to understand how this relates to the indexing RFCs. Fine to just handle DAG root to start but having w3s.link ONLY work for DAG root is very frustrating and I see the indexing work heading in a direction where we're not limited in that way.